afqinsight/CovariateRegressor.py

```diff
@@ -0,0 +1,225 @@
+
+
+def find_subset_indices(X_full, X_subset, method="hash"):
+    """
+    Find row indices in X_full that correspond to rows in X_subset.
+    Supports 'hash' (fast) and 'precise' (element-wise) matching.
+    """
+    if X_full.shape[1] != X_subset.shape[1]:
+        raise ValueError(
+            f"Feature dimensions don't match: {X_full.shape[1]} vs {X_subset.shape[1]}"
+        )
+    indices = []
+    if method == "precise":
+        for i, subset_row in enumerate(X_subset):
+            matches = [
+                j
+                for j, full_row in enumerate(X_full)
+                if np.array_equal(full_row, subset_row, equal_nan=True)
+            ]
+            if not matches:
+                raise ValueError(f"No matching row found for subset row {i}")
+            indices.append(matches[0])
+    elif method == "hash":
+        full_hashes = [hash(row.tobytes()) for row in X_full]
+        for i, subset_row in enumerate(X_subset):
+            subset_hash = hash(subset_row.tobytes())
+            try:
+                indices.append(full_hashes.index(subset_hash))
+            except ValueError as e:
+                raise ValueError(f"No matching row found for subset row {i}") from e
+    else:
+        raise ValueError(f"Unknown method '{method}'. Use 'hash' or 'precise'.")
+    return np.array(indices)
+
+
+class IdentityTransformer(BaseEstimator, TransformerMixin):
+    """A transformer that returns the input unchanged."""
+
+    def fit(self, X, y=None):
+        return self
+
+    def transform(self, X):
+        return X
+
+
+class CovariateRegressor(BaseEstimator, TransformerMixin):
+    """
+    Fits covariate(s) onto each feature in X and returns their residuals.
+    """
+
+    def __init__(
+        self,
+        covariate,
+        X,
+        pipeline=None,
+        cross_validate=True,
+        precise=False,
+        unique_id_col_index=None,
+        stack_intercept=True,
+    ):
+        """Regresses out a variable (covariate) from each feature in X.
+
+        Parameters
+        ----------
+        covariate : numpy array
+            Array of length (n_samples, n_covariates) to regress out of each
+            feature; May have multiple columns for multiple covariates.
+        X : numpy array
+            Array of length (n_samples, n_features), from which the covariate
+            will be regressed. This is used to determine how the
+            covariate-models should be cross-validated (which is necessary
+            to use in in scikit-learn Pipelines).
+        cross_validate : bool
+            Whether to cross-validate the covariate-parameters (y~covariate)
+            estimated from the train-set to the test set (cross_validate=True)
+            or whether to fit the covariate regressor separately on the test-set
+            (cross_validate=False). Setting this parameter to True is equivalent
+            to "foldwise covariate regression" (FwCR) as described in our paper
```

Comment:
```suggestion
            to "foldwise covariate regression" (FwCR) as described in Snoek et al. 
```

And add a reference in the Notes section. 

BTW: if you used code from Snoek et al.'s previous work that's probably fine, but we might need to add their original license somewhere, to comply with that license's requirements. Do you have a pointer to the original code?

Reply:
https://github.com/lukassnoek/MVCA?tab=MIT-1-ov-file#readme

This is the original repository and the original license is a MIT one

Reply:
Gotcha! As you can see, the license states that: 

```
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
```

I would add that either in our license file. Something like: 

```
Code used in the covariate_regressor module is distributed with the following license: 
```

And then you past their entire license after it.
 
BTW: I would also change the file-name to `covariate_regressor.py`, because it is not usual to have Python module names with CamelCase.

Reply:
Or maybe `covariate_regression.py`?

Reply:
Renamed files and appended suggestion to AFQ-Insight’s license file below our BSD-3 license